platform/engineering/design-docs/2020-12-10-auto-deploy-cms-build.md

```diff
@@ -0,0 +1,216 @@
+
+The schedule to start with will be hourly between 8am and 8pm on weekdays. Ideally, the deploy would run every 5 minutes.
+
+```
+node('vagov-autodeploy') {
+  triggers {
+    cron('0 8-21 * * 1-5')
+  }
+
+  dir('vets-website') {
+    ...
+  }
+
+  stage('Build') {
+    ...
+  }
+}
+```
+
+#### Preventing deploys of existing content
+
+On successful deploys, the tarball containing the CMS export data used in the build should be archived in S3. A possible key might be `s3://vetsgov-website-builds-s3-upload/cms-export/data/latest.tar.bz2`.
+
+The CMS export build [currently reads the tarball directly from the response body](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/drupal/api.js#L147) and should be modified to save it to a file.
+
+Then on subsequent deploys, the tarball downloaded from Drupal will be compared with the tarball of the previous successful deploy. The build will run only if the MD5 checksums of the two tarballs differ. This ensures that we only run the build when content has changed.
+
+For convenience, we could also upload the checksum to `s3://vetsgov-website-builds-s3-upload/cms-export/data/checksum.txt`. With that, the pipeline would not have to download the full tarball for the comparison.
+
+### Code Location
+
+The pipeline will be defined in `Jenkinsfile.autodeploy`. The [helper functions](https://github.com/department-of-veterans-affairs/vets-website/blob/master/jenkins/common.groovy) that it uses are in `jenkins/common.groovy`.
+
+There may be auxiliary configurations required in the `devops` [repo](https://github.com/department-of-veterans-affairs/devops), likely in the `ansible` directory.
+
+### Testing Plan
+
+The following behaviors should be observed cumulatively as they are implemented.
+1. Build is stored at the proper location in S3 when the pipeline is invoked.
+2. Pipeline runs on the defined schedule.
+3. Deploy doesn't proceed when no new content has been published.
+
+### Logging
+
+On top of the existing build logs, the new content-only deploy will output logs for the new transformation processes in the CMS export build.
+
+### Debugging
+
+The auto-deploy pipeline can be debugged by reading its console output. The S3 bucket can be checked to confirm whether the build has been deployed.
+
+The CMS export build will validate the input and output schemas of the transformers, so errors related to the build itself can be debugged using that output.
+
+As with the full build, during the rollout, there will be a comparison of GraphQL and CMS export builds to report errors when discrepancies between the two are found. The script used for this comparison will have some console output as well as an output JSON file describing the discrepancies.
+
+To determine whether a page exists in the Drupal content, we can go to [`staging.va.gov/drupal/debug/`](staging.va.gov/drupal/debug/). This page is not available in production, however.
+
```

Comment:
Additional concerns that ought to be baked in:
1.  Will content releases fail on broken link checking?  (please no  https://github.com/department-of-veterans-affairs/va.gov-team/pull/14092)   
2.  In the event of the failure, due to broken link of some other content issued that prevents a successful build.... what will happen?  (ex: Queue for another attempt, just wait until the next timed release,...?)